internal/ingress/controller/controller.go

```diff
@@ -668,6 +680,13 @@ func (n *NGINXController) createUpstreams(data []*extensions.Ingress, du *ingres
 				}
 			}
 
+			// marks the upstream as virtual for association with real upstream
+			if anns.Canary.Enabled {
+				upstreams[defBackend].Virtual = true
```

Comment:
You only call mergeVirtualBackends when anns.Canary.Enabled is true. Are these markings necessary? It seems like a redundant check.

Reply:
Setting the `Virtual` field to true on the backend is necessary since its used in the merge function here:
https://github.com/Shopify/ingress/blob/abd198fc93825399648a659bccec6b77ecd3cc7b/internal/ingress/controller/controller.go#L1142

My motivation here was to not couple the idea of canary too tightly with concept of virtual backends to stay flexible at the cost of some redundancy.